### PR TITLE
Fix table of contents for headings in tables and collapsible sections

### PR DESCRIPTION
**Pull request by KatsiarynaDzibrova**
Fixes following issues with table of contents:
- The order in table of contents is wrong for heading inside tables and collapsible sections.
- The heading is duplicated when resizing parent table cell

Now on each heading update dfs is used to get all headings and create a new TOC from scratch. It might take more time, but the change shouldn't be noticeable

Before:
toc-before

After:
toc-after

### CODE REVIEW

File: packages/lexical-react/src/LexicalTableOfContents.tsx [hunk 1]
```diff
@@ -138,42 +86,15 @@ export default function LexicalTableOfContentsPlugin({
...
-                );
-              }
-            }
-          }
+          const dfsNodes = $dfs().filter((dfsNode: DFSNode) =>
+            $isHeadingNode(dfsNode.node),
+          );
+          const headingNodes = dfsNodes.map(
+            (dfsNode) => dfsNode.node as HeadingNode,
+          );
```
Comment (ivailop7) [inline packages/lexical-react/src/LexicalTableOfContents.tsx:96]: You can reduce from 2 loops to 1 loop using reduce or flatmap. Syntax should be something along those lines:

```typescript
const headingNodes = $dfs().reduce((headingNodesList, node) => $isHeadingNode(dfsNode.node) ? [...acc, value] : acc, [])
```

**Review by ivailop7 — COMMENTED** [review]
A small speedup comment from me. @zurfyx @acy, happy with this approach?

File: packages/lexical-react/src/LexicalTableOfContents.tsx (hunk 1 above)
Comment (KatsiarynaDzibrova) [reply packages/lexical-react/src/LexicalTableOfContents.tsx:94]: Thank you. Added the change

**Review by ivailop7 — DISMISSED** [review]

File: packages/lexical-react/src/LexicalTableOfContents.tsx [hunk 2]
```diff
@@ -138,42 +86,13 @@ export default function LexicalTableOfContentsPlugin({
...
-                  currentTableOfContents,
-                );
-              }
-            }
-          }
+          const headingNodes: HeadingNode[] = $dfs().reduce(
+            (acc, {node}) => ($isHeadingNode(node) ? [...acc, node] : acc),
+            [] as HeadingNode[],
+          );
+          currentTableOfContents = $newTableOfContents(headingNodes);
```
Comment (ivailop7) [inline packages/lexical-react/src/LexicalTableOfContents.tsx:95]: I think we can do an extra performance optimization - we store the list, and we only replace it with the output from dfs on create, update for a headingNode, but on delete, we just delete from the map. What do you thing?

**Review by zurfyx — COMMENTED** [review]
Have we measured performance? Running a full `$dfs` every time that a HeadingNode is mutated might be fairly inefficient, that why the previous version uses mutated nodes. For reference, the reconciler doesn't look at all nodes either, instead it traverses through the dirty nodes, I would take the reconciler as an example for optimizing this plugin.

**Comment by KatsiarynaDzibrova** [comment]
Made the following changes:
1. Explicitly checking if the heading is already in TOC to avoid duplicates
2. Replaced dfs with listener on root updates, which updates all child headings
3. Find previous heading for a new heading only when it's updated
The performance should be much better now as we only visit updated nodes

@zurfyx feel free to take a look

File: packages/lexical-utils/src/index.ts [hunk 3]
```diff
@@ -222,6 +222,29 @@ function $getDepth(node: LexicalNode): number {
   return depth;
 }
 
+export function $getPreviousNode(
```
Comment (ivailop7) [inline packages/lexical-utils/src/index.ts:225]: It lacks a comment description, also confused by the naming getPreviousNode, yet in the first if statement, it gets the last child, which is not 'previous' to my understanding.

Comment (ivailop7) [reply packages/lexical-utils/src/index.ts:225]: Still unsure if this should be in utils or encapsulated in the TOC plugin file. Thoughts @zurfyx ?

**Review by zurfyx — APPROVED** [review]

**Pull request merged**